modules/ml/src/main/java/org/apache/ignite/ml/inference/builder/IgniteDistributedInfModelBuilder.java

```diff
@@ -0,0 +1,367 @@
+
+    /** Default configuration for Apache Ignite queues used in this class (request queue, response queue). */
+    private static final CollectionConfiguration queueCfg = new CollectionConfiguration();
+
+    /** Ignite instance. */
+    private final Ignite ignite;
+
+    /** Number of service instances maintaining to make distributed inference. */
+    private final int instances;
+
+    /** Max per node number of instances. */
+    private final int maxPerNode;
+
+    /**
+     * Constructs a new instance of Ignite distributed inference model builder.
+     *
+     * @param ignite Ignite instance.
+     * @param instances Number of service instances maintaining to make distributed inference.
+     * @param maxPerNode Max per node number of instances.
+     */
+    public IgniteDistributedInfModelBuilder(Ignite ignite, int instances, int maxPerNode) {
+        this.ignite = ignite;
+        this.instances = instances;
+        this.maxPerNode = maxPerNode;
+    }
+
+    /**
+     * Starts the specified in constructor number of service instances and request/response queues. Each service
+     * instance reads request queue, processes inbound requests and writes responses to response queue. The returned
+     * facade is represented by the {@link InfModel} operates with request/response queues, but hides these details
+     * behind {@link InfModel#predict(Object)} method of {@link InfModel}.
+     *
+     * Be aware that {@link InfModel#close()} method must be called to clear allocated resources, stop services and
+     * remove queues.
+     *
+     * @param reader Inference model reader.
+     * @param parser Inference model parser.
+     * @param <I> Type of model input.
+     * @param <O> Type of model output.
+     * @return Facade represented by {@link InfModel}.
+     */
+    @Override public <I extends Serializable, O extends Serializable> InfModel<I, Future<O>> build(
+        InfModelReader reader, InfModelParser<I, O> parser) {
+        return new DistributedInfModel<>(ignite, UUID.randomUUID().toString(), reader, parser, instances, maxPerNode);
+    }
+
+    /**
+     * Facade that operates with request/response queues to make distributed inference, but hides these details
+     * behind {@link InfModel#predict(Object)} method of {@link InfModel}.
+     *
+     * Be aware that {@link InfModel#close()} method must be called to clear allocated resources, stop services and
+     * remove queues.
+     *
+     * @param <I> Type of model input.
+     * @param <O> Type of model output.
+     */
+    private static class DistributedInfModel<I extends Serializable, O extends Serializable>
+        implements InfModel<I, Future<O>> {
+        /** Ignite instance. */
+        private final Ignite ignite;
+
+        /** Suffix that with correspondent templates formats service and queue names. */
+        private final String suffix;
+
+        /** Request queue. */
+        private final IgniteQueue<I> reqQueue;
```

Comment:
Are you really here need an IgniteQueue here instead of j.u.c primitives, could you please explain?